ingestion-edge/ingestion_edge/util.py

```diff
@@ -4,41 +4,102 @@
 
 """Utilities."""
 
+from google.cloud.pubsub_v1.gapic.publisher_client import PublisherClient
+from google.cloud.pubsub_v1.publisher.exceptions import PublishError
+from google.cloud.pubsub_v1.types import BatchSettings, PublishRequest, PubsubMessage
+from typing import List, Optional
 import asyncio
 
 
-def async_wrap(future, loop=None):
-    """Wrap a threading-like future in an async future that can be awaited.
+class AsyncioBatch:
+    """Batch for google.cloud.pubsub_v1.PublisherClient in asyncio.
 
-    example:
-
-    async def func():
-        future = ...
-        await wrap(future)
+    Work around for https://github.com/googleapis/google-cloud-python/issues/7104
     """
-    async_future = asyncio.Future()
 
-    def result():
-        try:
-            async_future.set_result(future.result())
-        except Exception as e:
-            async_future.set_exception(e)
+    def __init__(
+        self,
+        client: PublisherClient,
+        topic: str,
+        settings: BatchSettings,
+        autocommit: bool = True,
+    ):
+        """Initialize."""
+        self.client = client
+        self.topic = topic
+        self.settings = settings
+
+        self.full = asyncio.Event()
+        self.messages: List[PubsubMessage] = []
+        # fix https://github.com/googleapis/google-cloud-python/issues/7108
+        self.size = PublishRequest(topic=topic, messages=[]).ByteSize()
+
+        # Create a task to commit when full
+        self.result = asyncio.create_task(self.commit())
+
+        # If max latency is specified start a task to monitor the batch
+        # and commit when max latency is reached
+        if autocommit and self.settings.max_latency < float("inf"):
+            asyncio.create_task(self.monitor())
+
+    async def monitor(self):
+        """Sleep until max latency is reached then set the batch to full."""
+        await asyncio.sleep(self.settings.max_latency)
+        self.full.set()
+
+    async def commit(self) -> List[str]:
+        """Publish this batch when full."""
+        await self.full.wait()
+
+        if not self.messages:
+            return []
+
+        response = await asyncio.get_running_loop().run_in_executor(
+            None, self.client.api.publish, self.topic, self.messages
+        )
+
+        if len(response.message_ids) != len(self.messages):
+            raise PublishError("Some messages not successfully published")
+
+        return response.message_ids
```

Comment:
I'm still wrapping my head around much of python's async features, so reading this is a good learning experience. I like the way this reads.

Is there any need to cancel tasks once we commit? If `full` is set due to hitting max size, what happens to the `monitor` task? Does it still sleep and then set `full` again? Does that prevent this class instance from getting cleaned up?

Reply:
> Is there any need to cancel tasks once we commit?

I tried to do that in flush, it introduced a weird bug that i couldn't figure out how to fix where it propagated down the stack to `run_in_exector` and raised `concurrent.futures.CancelledError` instead of `asyncio.CancelledError`, and that would in turn kill `uvloop`, instead of raising back up the stack.

> If `full` is set due to hitting max size, what happens to the `monitor` task? Does it still sleep and then set `full` again?

yes, but setting `full` again is a noop because the event is already set.

Note that `asyncio.Event` stays set and waiting on it after that is a noop, where `asyncio.Condition` will only notify pending tasks and waiting on it after that waits for the next notify. This is the exact opposite of what I expect from the names, but it's been years since I last dealt with concurrency primitives.

> Does that prevent this class instance from getting cleaned up?

nope, the task will end, the other tasks will already all be complete, there will be no references, and it will be cleaned up.